homeassistant/components/wemo/entity.py

```diff
@@ -170,19 +185,38 @@ async def async_added_to_hass(self) -> None:
         registry = self.hass.data[WEMO_DOMAIN]["registry"]
         await self.hass.async_add_executor_job(registry.register, self.wemo)
         registry.on(self.wemo, None, self._subscription_callback)
+        if self.wemo.supports_long_press():
+            try:
+                await self.hass.async_add_executor_job(
+                    self.wemo.ensure_long_press_virtual_device
+                )
+            except (AttributeError, ActionException):
+                _LOGGER.exception(
+                    "Failed to enable long press support for %s", self.name
+                )
 
     async def async_will_remove_from_hass(self) -> None:
         """Wemo device removed from hass."""
         registry = self.hass.data[WEMO_DOMAIN]["registry"]
         await self.hass.async_add_executor_job(registry.unregister, self.wemo)
 
     def _subscription_callback(
-        self, _device: WeMoDevice, _type: str, _params: str
+        self, _device: WeMoDevice, event_type: str, _params: str
     ) -> None:
         """Update the state by the Wemo device."""
-        _LOGGER.info("Subscription update for %s", self.name)
-        updated = self.wemo.subscription_update(_type, _params)
-        self.hass.add_job(self._async_locked_subscription_callback(not updated))
+        _LOGGER.info("Subscription event (%s) for %s", event_type, self.name)
+        if event_type == EVENT_TYPE_LONG_PRESS:
+            self.hass.bus.fire(
```

Comment:
Entities should normally not fire custom events as they already spawn state changed events. Events should be fired outside of platforms from the integration. See eg the deconz integration.

https://github.com/home-assistant/core/blob/2555b10d49371f1b408fc3701acfcf106e0dfe73/homeassistant/components/deconz/deconz_event.py#L53

Reply:
If it's a stateless device, like a push button, we want it to be represented by custom events. But they should not be fired from an entity.

Reply:
The event logic has been moved to a new file, `wemo_device.py`